### PR TITLE
Preview - Keycloak Test Framework

### PR DESCRIPTION
**Pull request by lhanusov**
Closes [#32399](https://github.com/keycloak/keycloak/issues/32399)

### CODE REVIEW

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+== Brighter Future?
+
+Keycloak team had started a https://github.com/orgs/keycloak/projects/42[joint effort] between the engineering and quality engineering people to design a new https://github.com/keycloak/keycloak/tree/main/test-framework[test framework] (May 2024). It had been initiated with a prototype to verify, if our thoughts are feasible. The prototype should have been based on the _JUnit5_ testing framework, specifically to implement _JUnit5_ callback classes which extend the default test lifecycle functionality and provide custom inject annotations, like _@KeycloakIntegrationTest_, _@InjectWebDriver_ or _@InjectRealm_.
```
Comment (vaceksimon) [inline blog/2024/preview-keycloak-test-framework.adoc:14]:
```suggestion
Keycloak team had started a https://github.com/orgs/keycloak/projects/42[joint effort] between the engineering and quality engineering people to design a new test framework (May 2024). It had been initiated with a prototype to verify if our thoughts were feasible. The prototype was based on the _JUnit5_ testing framework, specifically to implement _JUnit5_ callback classes which extend the default test lifecycle functionality and provide custom inject annotations, like _@KeycloakIntegrationTest_, _@InjectWebDriver_ or _@InjectRealm_.
```

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+For our purpose we've https://github.com/keycloak/keycloak/blob/main/test-framework/src/main/java/org/keycloak/test/framework/KeycloakIntegrationTestExtension.java[implemented] these three callback classes:
```
Comment (vaceksimon) [inline blog/2024/preview-keycloak-test-framework.adoc:16]:
```suggestion
For our purpose, we have https://github.com/keycloak/keycloak/blob/main/test-framework/src/main/java/org/keycloak/test/framework/KeycloakIntegrationTestExtension.java[implemented] these three callback classes:
```

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+----
+BeforeEachCallback
+AfterEachCallback
+AfterAllCallback
+----
+
+After a successful test round, we've continued with a proof of concept extending features list to support multiple server modes, different databases and _WebDrivers_, clients and users setup, _SmallRye_ configuration support, _OAuthClient_ based on https://connect2id.com/products/nimbus-oauth-openid-connect-sdk[_Nimbus SDK_] (this feature is a preview only) etc. The full list of currently implemented features can be found here in the https://github.com/keycloak/keycloak/tree/main/test-framework/src/main/java/org/keycloak/test/framework/annotations[annotations package].
+
+Are you curious where to start?
```
Comment (vaceksimon) [inline blog/2024/preview-keycloak-test-framework.adoc:25]:
```suggestion
== Are you curious about where to start?
```

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+We suggest to read the https://github.com/keycloak/keycloak/blob/main/test-framework/README.md[user guide], which will provide a basic overview how the framework works and should be used. If this is not enough, you can also check https://github.com/keycloak/keycloak/tree/main/test-poc/examples[examples module].
```
Comment (vaceksimon) [inline blog/2024/preview-keycloak-test-framework.adoc:26]:
```suggestion
We suggest reading the https://github.com/keycloak/keycloak/blob/main/test-framework/README.md[user guide], which will provide a basic overview of how the framework works and should be used. If this is not enough, you can also check https://github.com/keycloak/keycloak/tree/main/test-poc/examples[examples module].
```

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+Once you find some bug or want to discuss something, or propose a new enhancement, please follow this GitHub feedback https://www.github.com/discussion#blabla[discussion] link.
```
Comment (vaceksimon) [inline blog/2024/preview-keycloak-test-framework.adoc:28]:
```suggestion
Once you find a bug, want to discuss something, or propose a new enhancement, please follow this GitHub feedback https://www.github.com/discussion#blabla[discussion] link.
```

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+== Coming Soon
+
+Preview - https://github.com/keycloak/keycloak/tree/main/test-framework[_Keycloak Test Framework_] will be introduced in _Keycloak 26_. We plan to add more features, which will cover 80% of functionalities currently available in the https://github.com/keycloak/keycloak/tree/main/testsuite[integration test suite]. We also plan to introduce some new concepts, e.g. _@KeycloakTest_ annotation, which will add an option to test providers without _Keycloak_ server running. It should be based on the _KeycloakSession_ class. Another interesting features should be parallel execution on a GitHub runner, adding custom extension to a Keycloak server or enable Keycloak cluster setup.
```
Comment (vaceksimon) [inline blog/2024/preview-keycloak-test-framework.adoc:32]:
```suggestion
Preview - https://github.com/keycloak/keycloak/tree/main/test-framework[_Keycloak Test Framework_] will be introduced in _Keycloak 26_. We plan to add more features, which will cover 80% of functionalities currently available in the https://github.com/keycloak/keycloak/tree/main/testsuite[integration test suite]. We also plan to introduce some new concepts, e.g. _@KeycloakTest_ annotation, which will add an option to test providers without the _Keycloak_ server running. It should be based on the _KeycloakSession_ class. Other interesting features should be parallel execution on a GitHub runner, adding a custom extension to a Keycloak server, or enabling a Keycloak cluster setup.
```

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+== Acknowledgement
+
+I would like to thanks to all people who put the proof of concept together and made it a real: Miquel, Simon, Filip, Moises, Jon, Pedro. Special thanks goes to Stian, who led the technical design and proposed very nifty things that raised the project to another level.
```
Comment (vaceksimon) [inline blog/2024/preview-keycloak-test-framework.adoc:36]:
```suggestion
I would like to thank all the people who put the proof of concept together and made it real: Miquel, Simon, Filip, Moises, Jon, and Pedro. A special thank you goes to Stian, who led the technical design and proposed very nifty things that raised the project to another level.
```

**Review by vaceksimon — CHANGES_REQUESTED** [review]
Content-wise I don't find it missing anything. It's great as it is now.
I ran it through a grammar checker and included some of my rephrasing suggestions.
Feel free to ignore any of it.

**Review by vaceksimon — APPROVED** [review]

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+AfterEachCallback
+AfterAllCallback
+----
+
+After a successful test round, we've continued with a proof of concept extending features list to support multiple server modes, different databases and _WebDrivers_, clients and users setup, _SmallRye_ configuration support, _OAuthClient_ based on https://connect2id.com/products/nimbus-oauth-openid-connect-sdk[_Nimbus SDK_] (this feature is a preview only) etc. The full list of currently implemented features can be found here in the https://github.com/keycloak/keycloak/tree/main/test-framework/src/main/java/org/keycloak/test/framework/annotations[annotations package].
+
+== Are you curious about where to start?
+We suggest reading the https://github.com/keycloak/keycloak/blob/main/test-framework/README.md[user guide], which will provide a basic overview of how the framework works and should be used. If this is not enough, you can also check https://github.com/keycloak/keycloak/tree/main/test-poc/examples[examples module].
+
+For extension developers we recommend to look into the showcase example on how to start _Keycloak_ with their custom provider - link.
```
Comment (stianst) [inline blog/2024/preview-keycloak-test-framework.adoc:28]: Just a note that we need to update this before publishing the blog

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+Once you find a bug, want to discuss something, or propose a new enhancement, please follow this GitHub feedback https://www.github.com/discussion#blabla[discussion] link.
```
Comment (stianst) [inline blog/2024/preview-keycloak-test-framework.adoc:30]: Just a note that we need to update this before publishing the blog

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+== Coming Soon
+
+Preview - https://github.com/keycloak/keycloak/tree/main/test-framework[_Keycloak Test Framework_] will be introduced in _Keycloak Nightly_ builds. We plan to add more features, which will cover 80% of functionalities currently available in the https://github.com/keycloak/keycloak/tree/main/testsuite[integration test suite]. We also plan to introduce some new concepts, e.g. _@KeycloakTest_ annotation, which will add an option to test providers without the _Keycloak_ server running. It should be based on the _KeycloakSession_ class. Other interesting features should be parallel execution on a GitHub runner, adding a custom extension to a Keycloak server, or enabling a Keycloak cluster setup.
```
Comment (stianst) [inline blog/2024/preview-keycloak-test-framework.adoc:34]: Is available, not will be ;)

Comment (stianst) [reply blog/2024/preview-keycloak-test-framework.adoc:34]: Remove 80%, and say something like cover enough for most tests...

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+== Brighter Future?
+
+Keycloak team had started a https://github.com/orgs/keycloak/projects/42[joint effort] between the engineering and quality engineering people to design a new test framework (May 2024). It had been initiated with a prototype to verify if our thoughts were feasible. The prototype was based on the _JUnit5_ testing framework, specifically to implement _JUnit5_ callback classes which extend the default test lifecycle functionality and provide custom inject annotations, like _@KeycloakIntegrationTest_, _@InjectWebDriver_ or _@InjectRealm_.
```
Comment (stianst) [inline blog/2024/preview-keycloak-test-framework.adoc:14]: The board is private, and not visible to everyone. So we either remove the link, or cleanup the board and make it public

File: blog/2024/preview-keycloak-test-framework.adoc (hunk 9 above)
Comment (stianst) [inline blog/2024/preview-keycloak-test-framework.adoc:30]:
```suggestion
If you find a bug, want to discuss something, or propose a new enhancement, please follow this GitHub feedback https://www.github.com/discussion#blabla[discussion] link.
```

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+It would be good to mention a few drawbacks, that stand out when working with the current https://github.com/keycloak/keycloak/tree/main/testsuite[test suite]. First of all, is the complexity of various configurations and additions made on top of the https://arquillian.org/[_Arquillian framework_]. These changes make the test suite powerful, but the cons is that without proper documentation for beginners is almost unreadable. The second thing has the same importance, the _Arquillian framework_ is not fully supported anymore. Other things to mention are a complicated execution system, where you want to specify what exactly should be tested, then abstract classes with shared configurations, missing the option to add a custom extension, and last but not least, the massive amount of tests without knowledge if they are needed anymore.
```
Comment (stianst) [inline blog/2024/preview-keycloak-test-framework.adoc:10]: Not convinced about the statement `the massive amount of tests without knowledge if they are needed anymore`. There may be some tests not needed, but probably not a massive amount. Perhaps just drop this?

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+== Brighter Future?
+
+The Keycloak team began an effort to design a new test framework in May 2024. It started with a prototype to verify if our ideas were feasible. The prototype was based on the https://junit.org/junit5/[_JUnit5_ testing framework], specifically to implement https://junit.org/junit5/docs/current/user-guide/#extensions-lifecycle-callbacks[_JUnit5_ callback classes] which extend the default test lifecycle functionality and provide custom inject annotations, like _@KeycloakIntegrationTest_, _@InjectWebDriver_ or _@InjectRealm_.
```
Comment (stianst) [inline blog/2024/preview-keycloak-test-framework.adoc:14]:
```suggestion
The Keycloak team began an effort to design a new test framework in May 2024. It started with a prototype to verify if our ideas were feasible. The prototype was based on the https://junit.org/junit5/[_JUnit5_ testing framework], specifically to implement https://junit.org/junit5/docs/current/user-guide/#extensions-lifecycle-callbacks[_JUnit5_ callback classes] which extend the default test lifecycle functionality and provide custom inject annotations, like _@KeycloakIntegrationTest_, _@InjectWebDriver_ or _@InjectRealm_.
```

Perhaps writing something like `The prototype is an JUnit 5 Extension...`

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+For our purpose, we have https://github.com/keycloak/keycloak/blob/main/test-framework/core/src/main/java/org/keycloak/test/framework/KeycloakIntegrationTestExtension.java[implemented] these three callback classes:
+----
+BeforeEachCallback
+AfterEachCallback
+AfterAllCallback
+----
```
Comment (stianst) [inline blog/2024/preview-keycloak-test-framework.adoc:21]: Would drop this, folks just need to know it's a JUnit 5 extension that is automatically added when you add the dependency; they don't need to know the internals.

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+After a successful test round, we've continued with a proof of concept extending features list to support multiple server modes, different databases and _WebDrivers_, clients and users setup, _SmallRye_ configuration support, _OAuthClient_ based on https://connect2id.com/products/nimbus-oauth-openid-connect-sdk[_Nimbus SDK_] (this feature is a preview only) etc. The full list of currently implemented features can be found here in the https://github.com/keycloak/keycloak/tree/main/test-framework/core/src/main/java/org/keycloak/test/framework/annotations[annotations package]. It is already present in the main branch and _Keycloak_ nightly builds.
```
Comment (stianst) [inline blog/2024/preview-keycloak-test-framework.adoc:23]: The full list of features in the annotations package isn't the best way to list the features; and now that we split it up into multiple modules it doesn't list all annotations.

Comment (stianst) [reply blog/2024/preview-keycloak-test-framework.adoc:23]: Maybe just do a bullet point list of features?

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+If you find a bug, want to discuss something, or propose a new enhancement, please follow this GitHub feedback https://www.github.com/discussion#blabla[discussion link].
+
+== Coming Soon
+
+Main plan is to cover enough features for most functionalities currently available in the https://github.com/keycloak/keycloak/tree/main/testsuite[integration test suite]. Next future step shall be to introduce some new concepts, e.g. _@KeycloakTest_ annotation, which will add an option to test providers without the _Keycloak_ server running. It should be based on the _KeycloakSession_ class. Other interesting features should be https://junit.org/junit5/docs/current/user-guide/#junit-platform-suite-engine[_JUnit5_ test suites], parallel execution on a GitHub runner or enabling a Keycloak cluster setup.
```
Comment (stianst) [inline blog/2024/preview-keycloak-test-framework.adoc:34]:
```suggestion
== Next steps

We already have enough capabilities in the new test framework to start migrating some tests from the old testsuite; and in fact already have our very first test migrated. We plan to migrate one package at a time from the old testsuite starting with `admin` tests, then moving on to `forms` and `oauth` packages. As we are doing this we will expand on the capabilities of the test framework.

Some features we know will be coming soon include:

* An easier way to deploy custom providers, not requiring a Maven build of the provider first
* Improved logging, making it easy to configure logging from tests as well as Keycloak
* Easy testing of OAuth and OpenID Connect, including a mock application
* Extension to allow running code on the tested server when it's not possible to easily test through only remote interfaces
```

We also have some more long term plans to deliver:

* Provider tests that can be used to easily test a provider by invoking the provider directly
* Parallel execution of tests, to take full advantage of multiple cores to reduce test execution time

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 17]
```diff
@@ -0,0 +1,67 @@
+:title: "Writing new tests has never been easier than before." Keycloak Test Framework
```
Comment (stianst) [inline blog/2024/preview-keycloak-test-framework.adoc:1]: This title won't show very nicely on the website or feeds

```suggestion
:title: Introducing the Keycloak Test Framework
```

File: blog/2024/preview-keycloak-test-framework.adoc [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+:date: 2024-11-14
+:publish: true
+:author: Lukas Hanusovsky, Stian Thorgersen
```
Comment (stianst) [inline blog/2024/preview-keycloak-test-framework.adoc:4]:
```suggestion
:author: Lukas Hanusovsky
```

**Review by stianst — APPROVED** [review]

**Pull request merged**